flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/sorted/state/BatchExecutionInternalTimeService.java

```diff
@@ -0,0 +1,133 @@
+
+/**
+ * An implementation of a {@link InternalTimerService} that manages timers with a single active key at a time.
+ * Can be used in a BATCH execution mode.
+ */
+public class BatchExecutionInternalTimeService<K, N> implements InternalTimerService<N> {
+
+	private final ProcessingTimeService processingTimeService;
+
+	/**
+	 * Processing time timers that are currently in-flight.
+	 */
+	private final KeyGroupedInternalPriorityQueue<TimerHeapInternalTimer<K, N>> processingTimeTimersQueue;
+
+	/**
+	 * Event time timers that are currently in-flight.
+	 */
+	private final KeyGroupedInternalPriorityQueue<TimerHeapInternalTimer<K, N>> eventTimeTimersQueue;
+
+	/**
+	 * The local event time, as denoted by the last received
+	 * {@link org.apache.flink.streaming.api.watermark.Watermark Watermark}.
+	 */
+	private long currentWatermark = Long.MIN_VALUE;
+
+	private final Triggerable<K, N> triggerTarget;
+
+	private K currentKey;
+
+	BatchExecutionInternalTimeService(
+			ProcessingTimeService processingTimeService,
+			Triggerable<K, N> triggerTarget) {
+
+		this.processingTimeService = checkNotNull(processingTimeService);
+		this.triggerTarget = checkNotNull(triggerTarget);
+
+		this.processingTimeTimersQueue = new BatchExecutionInternalPriorityQueueSet<>(
+			PriorityComparator.forPriorityComparableObjects(),
+			128
+		);
+		this.eventTimeTimersQueue = new BatchExecutionInternalPriorityQueueSet<>(
+			PriorityComparator.forPriorityComparableObjects(),
+			128
+		);
+	}
+
+	@Override
+	public long currentProcessingTime() {
+		return processingTimeService.getCurrentProcessingTime();
+	}
+
+	@Override
+	public long currentWatermark() {
+		return currentWatermark;
+	}
+
+	@Override
+	public void registerProcessingTimeTimer(N namespace, long time) {
+		processingTimeTimersQueue.add(new TimerHeapInternalTimer<>(time, currentKey, namespace));
+	}
+
+	@Override
+	public void registerEventTimeTimer(N namespace, long time) {
+		eventTimeTimersQueue.add(new TimerHeapInternalTimer<>(time, currentKey, namespace));
+	}
+
+	@Override
+	public void deleteProcessingTimeTimer(N namespace, long time) {
+		processingTimeTimersQueue.remove(new TimerHeapInternalTimer<>(time, currentKey, namespace));
+	}
+
+	@Override
+	public void deleteEventTimeTimer(N namespace, long time) {
+		eventTimeTimersQueue.remove(new TimerHeapInternalTimer<>(time, currentKey, namespace));
+	}
+
+	@Override
+	public void forEachEventTimeTimer(BiConsumerWithException<N, Long, Exception> consumer) {
+		throw new UnsupportedOperationException(
+			"The BatchExecutionInternalTimeService should not be used in State Processor API.");
+	}
+
+	@Override
+	public void forEachProcessingTimeTimer(BiConsumerWithException<N, Long, Exception> consumer) {
+		throw new UnsupportedOperationException(
+			"The BatchExecutionInternalTimeService should not be used in State Processor API.");
+	}
+
+	public void setCurrentKey(K currentKey) throws Exception {
```

Comment:
This relies on `setCurrentKey` only being called for key changes, not on every element, right?

Reply:
Yes. I will change it to check if the key changed. Good point.

Reply:
Had the same comment as @aljoscha plus, shouldn't we here prevent the user-code from registering any new timers? If not, it looks like we risk ending up in a loop where every fired timer will register a new one, right?
We can simply add in the register timer `if(watermark ==Long.MAX) -> ignore` or sth like that. Same for processing time (and I think it is ok to check again the watermark).

In addition, we are relying on each key being processed in its entirety and then move to the next, which is totally valid. In this case, I think we should somehow disable the `DataStreamUtils.reinterpretAsKeyedStream()` for batch because, if I am not mistaken, it messes things up, right? Maybe at another ticket @dawidwys ?

Reply:
As for the:
> If not, it looks like we risk ending up in a loop where every fired timer will register a new one, right?
We can simply add in the register timer if(watermark ==Long.MAX) -> ignore or sth like that. Same for processing time (and I think it is ok to check again the watermark).

If I am not mistaken that's also the current behaviour in a regular timer service. If you register a timer for `time < currentWatermark` you might end up in an infinite loop. Not sure if we should diverge the two implementations.

Reply:
I don't think `reinterpretAsKeyedStream` breaks anything. It just tells the stream is keyed. The sorting/grouping happens on the receiver side.

Reply:
What if it is the source "chain"? Which I think is the main usecase. "I have my Kafka input stream already grouped by key."

Reply:
As for the infinite loop, before we were "assuming" (maybe not correctly) that we are in an "unbounded stream" scenario, where there is no "MAX_WATERMARK". Of course this can happen at any point in time if we keep on registering timers for "earlier", but for that, there is no easy way of correcting it. Now, at the end of the stream, even if the user writes correct code that registers a timer to periodically execute something, this may lead to an infinite loop if the user-code does not check for MAX_WATERMARK.

Reply:
Good point with the `chaining`. I will keep that in mind when exposing this feature.

Reply:
After some more thinking, I agree with you @kl0u in regards to registering timers in the callback. Any such patterns that make sense in streaming, do not necessarily make sense in BATCH and will result in pausing the processing for a longer period of time.

I will quiesce the timer service as you suggested.